File: src/harness/measure.ts
import { EngineRequestTimeoutError, BenchmarkParseError } from "../errors.js";
import type { CompletionResult } from "../types.js";

export interface TimedCompletionOptions {
  engine: string;
  baseUrl: string;
  modelId: string;
  prompt: string;
  maxTokens: number;
  timeoutMs: number;
}

/**
 * Sends one timed chat-completion request to an engine's OpenAI-compatible
 * server and measures wall-clock time + reported token counts. This is the
 * ONE shared measurement code path for every engine -- the architecture
 * decision that replaced per-engine benchmark-CLI parsing (omlx has no CLI
 * benchmark tool at all; this makes the comparison genuinely apples-to-apples
 * since the same code measures every engine).
 */
export async function timedCompletion(
  opts: TimedCompletionOptions,
): Promise<CompletionResult> {
  const body = JSON.stringify({
    model: opts.modelId,
    messages: [{ role: "user", content: opts.prompt }],
    max_tokens: opts.maxTokens,
    stream: false,
  });

  // Elapsed time must be measured across the FULL request, including
  // reading the response body -- fetch() itself resolves as soon as
  // headers arrive, before generation is done streaming back. Measuring
  // right after `await fetch(...)` (a real bug caught during a live
  // end-to-end run: it produced a physically impossible 64,646 tok/s)
  // only captures time-to-first-byte, not actual generation time.
  const start = performance.now();
  let response: Response;
  try {
    response = await fetch(`${opts.baseUrl}/v1/chat/completions`, {
      method: "POST",
      headers: { "Content-Type": "application/json" },
      body,
      signal: AbortSignal.timeout(opts.timeoutMs),
    });
  } catch (err) {
    if (err instanceof Error && err.name === "TimeoutError") {
      throw new EngineRequestTimeoutError(opts.engine, opts.timeoutMs);
    }
    throw err;
  }

  if (!response.ok) {
    const errorBody = await response.text().catch(() => "");
    throw new BenchmarkParseError(opts.engine, `HTTP ${response.status}: ${errorBody}`);
  }

  let payload: unknown;
  const rawText = await response.text();
  const elapsedMs = performance.now() - start;
  try {
    payload = JSON.parse(rawText);
  } catch {
    throw new BenchmarkParseError(opts.engine, rawText);
  }

  const usage = (payload as Record<string, unknown>)?.usage as
    | Record<string, unknown>
    | undefined;
  const completionTokens =
    typeof usage?.completion_tokens === "number" ? usage.completion_tokens : null;
  const promptTokensDetails = usage?.prompt_tokens_details as
    | Record<string, unknown>
    | undefined;
  const cachedPromptTokens =
    typeof promptTokensDetails?.cached_tokens === "number"
      ? promptTokensDetails.cached_tokens
      : 0;

  const tokensPerSecond =
    completionTokens && elapsedMs > 0
      ? completionTokens / (elapsedMs / 1000)
      : null;

  return {
    elapsedMs: Math.round(elapsedMs),
    completionTokens,
    cachedPromptTokens,
    tokensPerSecond: tokensPerSecond ? Math.round(tokensPerSecond * 100) / 100 : null,
  };
}

File: python/src/inferbench/harness/measure.py
"""
Ported from src/harness/measure.ts. Sends one timed chat-completion request
to an engine's OpenAI-compatible server and measures wall-clock time plus
reported token counts -- the ONE shared measurement code path for every
engine (the architecture decision that replaced per-engine benchmark-CLI
parsing; omlx has no CLI benchmark tool at all, verified against its real
README before the original tool's adapter code was written).

Elapsed time is measured across the FULL response body, not just the
headers: urlopen() returns once headers arrive, before the response body
(and therefore generation) has finished streaming. The original TypeScript
harness had a real bug here once -- measuring right after `await fetch(...)`
resolved, which produced a physically impossible 64,646 tok/s during a live
end-to-end run before it was caught and fixed. This port measures after
`response.read()` completes, deliberately not reintroducing that bug.
"""
from __future__ import annotations

import json
import socket
import time
import urllib.error
import urllib.request
from dataclasses import dataclass

from ..errors import BenchmarkParseError, EngineRequestTimeoutError
from ..types import CompletionResult


@dataclass
class TimedCompletionOptions:
    engine: str
    base_url: str
    model_id: str
    prompt: str
    max_tokens: int
    timeout_ms: int


def timed_completion(opts: TimedCompletionOptions) -> CompletionResult:
    body = json.dumps(
        {
            "model": opts.model_id,
            "messages": [{"role": "user", "content": opts.prompt}],
            "max_tokens": opts.max_tokens,
            "stream": False,
        }
    ).encode("utf-8")

    request = urllib.request.Request(
        f"{opts.base_url}/v1/chat/completions",
        data=body,
        headers={"Content-Type": "application/json"},
        method="POST",
    )

    start = time.perf_counter()
    try:
        response = urllib.request.urlopen(  # noqa: S310 -- fixed, internally-constructed URL only
            request, timeout=opts.timeout_ms / 1000
        )
        raw_bytes = response.read()
    except urllib.error.HTTPError as err:
        error_body = err.read().decode("utf-8", errors="replace") if err.fp else ""
        raise BenchmarkParseError(opts.engine, f"HTTP {err.code}: {error_body}") from err
    except (socket.timeout, TimeoutError) as err:
        raise EngineRequestTimeoutError(opts.engine, opts.timeout_ms) from err
    except urllib.error.URLError as err:
        if isinstance(err.reason, (socket.timeout, TimeoutError)):
            raise EngineRequestTimeoutError(opts.engine, opts.timeout_ms) from err
        raise

    elapsed_ms = (time.perf_counter() - start) * 1000
    raw_text = raw_bytes.decode("utf-8", errors="replace")

    try:
        payload = json.loads(raw_text)
    except json.JSONDecodeError as err:
        raise BenchmarkParseError(opts.engine, raw_text) from err

    usage = payload.get("usage") if isinstance(payload, dict) else None
    completion_tokens = usage.get("completion_tokens") if isinstance(usage, dict) else None
    if not isinstance(completion_tokens, (int, float)) or isinstance(completion_tokens, bool):
        completion_tokens = None

    prompt_tokens_details = usage.get("prompt_tokens_details") if isinstance(usage, dict) else None
    cached_raw = (
        prompt_tokens_details.get("cached_tokens") if isinstance(prompt_tokens_details, dict) else None
    )
    cached_prompt_tokens = (
        int(cached_raw) if isinstance(cached_raw, (int, float)) and not isinstance(cached_raw, bool) else 0
    )

    tokens_per_second = (
        completion_tokens / (elapsed_ms / 1000) if completion_tokens and elapsed_ms > 0 else None
    )

    return CompletionResult(
        elapsed_ms=round(elapsed_ms),
        completion_tokens=int(completion_tokens) if completion_tokens is not None else None,
        cached_prompt_tokens=cached_prompt_tokens,
        tokens_per_second=round(tokens_per_second, 2) if tokens_per_second is not None else None,
    )
